File: http/src/test/scala/zipkin/finagle/http/HttpRawZipkinTracerTest.scala
package zipkin.finagle.http

import java.net.URI
import java.util
import java.util.Arrays.asList

import com.twitter.finagle.ChannelClosedException
import com.twitter.finagle.stats.InMemoryStatsReceiver
import com.twitter.finagle.tracing.{Flags, SpanId, TraceId}
import com.twitter.finagle.zipkin.core.{BinaryAnnotation, Endpoint, Span, ZipkinAnnotation}
import com.twitter.util.{Await, Base64StringEncoder, Time}
import org.junit.{ClassRule, Test}
import org.scalatest.junit.JUnitSuite
import zipkin.Codec
import zipkin.junit.{HttpFailure, ZipkinRule}

object HttpRawZipkinTracerTest {
  // Singleton as the test needs to read the actual port in use
  val zipkinRule = new ZipkinRule()

  // Scala cannot generate fields with public visibility, so use a def instead.
  @ClassRule def zipkinDef = zipkinRule
}

class HttpRawZipkinTracerTest extends JUnitSuite {

  import HttpRawZipkinTracerTest.zipkinRule

  val statsReceiver = new InMemoryStatsReceiver

  val traceId = TraceId(Some(SpanId(123)), Some(SpanId(123)), SpanId(123), None, Flags().setDebug)

  @Test def sendSpans {
    val localEndpoint = Endpoint(2323, 23)
    val remoteEndpoint = Endpoint(333, 22)

    val annotations = Seq(
      ZipkinAnnotation(Time.fromSeconds(123), "cs", localEndpoint),
      ZipkinAnnotation(Time.fromSeconds(126), "cr", localEndpoint),
      ZipkinAnnotation(Time.fromSeconds(123), "ss", remoteEndpoint),
      ZipkinAnnotation(Time.fromSeconds(124), "sr", remoteEndpoint),
      ZipkinAnnotation(Time.fromSeconds(123), "llamas", localEndpoint)
    )

    val span = Span(
      traceId = traceId,
      annotations = annotations,
      _serviceName = Some("hickupquail"),
      _name = Some("foo"),
      bAnnotations = Seq.empty[BinaryAnnotation],
      endpoint = localEndpoint)

    val expected = Base64StringEncoder.decode(
      """DAAAAAEKAAEAAAAAAAAAewsAAwAAAANmb28KAAQAAAAAAAAAewoABQAAAAAAAAB7DwAGDAA
        |AAAUKAAEAAAAAB1TUwAsAAgAAAAJjcwwAAwgAAQAACRMGAAIAFwsAAwAAAAtoaWNrdXBxdW
        |FpbAAACgABAAAAAAeCm4ALAAIAAAACY3IMAAMIAAEAAAkTBgACABcLAAMAAAALaGlja3Vwc
        |XVhaWwAAAoAAQAAAAAHVNTACwACAAAAAnNzDAADCAABAAABTQYAAgAWCwADAAAAC2hpY2t1
        |cHF1YWlsAAAKAAEAAAAAB2QXAAsAAgAAAAJzcgwAAwgAAQAAAU0GAAIAFgsAAwAAAAtoaWN
        |rdXBxdWFpbAAACgABAAAAAAdU1MALAAIAAAAGbGxhbWFzDAADCAABAAAJEwYAAgAXCwADAA
        |AAC2hpY2t1cHF1YWlsAAACAAkBAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
        |AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
        |AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
        |AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=""".stripMargin)

    Await.result(tracer().sendSpans(Seq(span)))

    assert(withoutTimestampAndDuration(zipkinRule.getTraces())
      == asList(Codec.THRIFT.readSpans(expected)))
    assert(statsReceiver.counters == Map(List("log_span", "ok") -> 1))
  }

  @Test def sendSpans_incrementsErrorCounterOnFailure {
    zipkinRule.enqueueFailure(HttpFailure.disconnectDuringBody())

    val span = Span(
      traceId = traceId,
      annotations = Seq.empty,
      _serviceName = Some("hickupquail"),
      _name = Some("foo"),
      bAnnotations = Seq.empty[BinaryAnnotation],
      endpoint = Endpoint(2323, 23))

    Await.result(tracer().sendSpans(Seq(span)))

    assert(statsReceiver.counters ==
      Map(List("log_span", "error",
        classOf[ChannelClosedException].getName()) -> 1))
  }

  def tracer(): HttpRawZipkinTracer = {
    val httpUrl = URI.create(zipkinRule.httpUrl())

    new HttpRawZipkinTracer(httpUrl.getHost + ":" + httpUrl.getPort, statsReceiver)
  }

  /** Finagle doesn't report the span's timestamp and duration, yet, as it hasn't updated thrifts */
  def withoutTimestampAndDuration(traces: util.List[util.List[zipkin.Span]]) = {
    import collection.JavaConverters._
    traces.asScala.map(spans => spans.asScala.map(s =>
      s.toBuilder().timestamp(null).duration(null).build()).asJava).asJava
  }
}

File: http/src/test/java/zipkin/finagle/http/HttpRawZipkinTracerTest.java
/**
 * Copyright 2016 The OpenZipkin Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package zipkin.finagle.http;

import com.twitter.finagle.stats.StatsReceiver;
import com.twitter.finagle.tracing.Annotation.ClientRecv;
import com.twitter.finagle.tracing.Annotation.ClientSend;
import com.twitter.finagle.tracing.Annotation.Rpc;
import com.twitter.finagle.tracing.Annotation.ServiceName;
import com.twitter.finagle.tracing.Record;
import com.twitter.util.Await;
import com.twitter.util.Duration;
import java.net.URI;
import java.util.List;
import org.junit.After;
import org.junit.Rule;
import org.junit.Test;
import scala.Option;
import zipkin.Span;
import zipkin.finagle.RawZipkinTracer;
import zipkin.finagle.RawZipkinTracerTest;
import zipkin.finagle.http.HttpZipkinTracer.Config;
import zipkin.junit.ZipkinRule;

import static com.twitter.util.Time.fromMilliseconds;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.entry;
import static scala.collection.JavaConversions.mapAsJavaMap;

public class HttpRawZipkinTracerTest extends RawZipkinTracerTest {

  @Rule
  public ZipkinRule http = new ZipkinRule();

  Config config =
      Config.builder().host("localhost:" + URI.create(http.httpUrl()).getPort()).build();

  @After
  public void closeClient() {
    ((HttpRawZipkinTracer) tracer).client.close();
  }

  @Override protected RawZipkinTracer newTracer(StatsReceiver stats) {
    return new HttpRawZipkinTracer(config.host(), stats);
  }

  @Override protected List<List<Span>> getTraces() {
    return http.getTraces();
  }

  @Test
  public void whenHttpIsDown() throws Exception {
    closeClient();
    config = Config.builder().host("127.0.0.1:65535").build();
    tracer = newTracer(stats);

    tracer.record(new Record(traceId, fromMilliseconds(TODAY), new ServiceName("web"), none));
    tracer.record(new Record(traceId, fromMilliseconds(TODAY), new Rpc("get"), none));
    tracer.record(new Record(traceId, fromMilliseconds(TODAY), new ClientSend(), none));
    tracer.record(new Record(traceId, fromMilliseconds(TODAY + 1), new ClientRecv(), none));

    Await.result(tracer.flush());

    assertThat(mapAsJavaMap(stats.counters())).containsExactly(
        entry(seq("log_span", "error", "com.twitter.finagle.ChannelWriteException"), 1)
    );
  }

  final Option<Duration> none = Option.empty(); // avoid having to force generics
}
